mat64/shadow.go

```diff
@@ -0,0 +1,201 @@
+		// We know m is completely after a.
+		return false
+	}
+
+	if mat.Stride != a.Stride {
+		// Too hard, so assume the worst.
+		panic(mismatchedStrides)
+	}
+
+	if off < 0 {
+		off = -off
+		mat.Cols, a.Cols = a.Cols, mat.Cols
+	}
+	if rectanglesOverlap(off, mat.Cols, a.Cols, mat.Stride) {
+		panic(regionOverlap)
+	}
+	return false
+}
+
+func (s *SymDense) checkOverlap(a blas64.Symmetric) bool {
+	mat := s.RawSymmetric()
+	if cap(mat.Data) == 0 || cap(a.Data) == 0 {
+		return false
+	}
+
+	off := offset(mat.Data[:1], a.Data[:1])
+
+	if off == 0 {
+		// At least one element overlaps.
+		if mat.N == a.N && mat.Stride == a.Stride {
+			panic(regionIdentity)
+		}
+		panic(regionOverlap)
+	}
+
+	if off > 0 && len(mat.Data) <= off {
+		// We know s is completely before a.
+		return false
+	}
+	if off < 0 && len(a.Data) <= -off {
+		// We know s is completely after a.
+		return false
+	}
+
+	if mat.Stride != a.Stride {
+		// Too hard, so assume the worst.
+		panic(mismatchedStrides)
+	}
+
+	// TODO(kortschak) Make this analysis more precise.
+	if off > 0 {
+		off = -off
+		mat.N, a.N = a.N, mat.N
+	}
+	if rectanglesOverlap(off, mat.N, a.N, mat.Stride) {
+		panic(regionOverlap)
+	}
+	return false
+}
+
+func (t *TriDense) checkOverlap(a blas64.Triangular) bool {
+	mat := t.RawTriangular()
+	if cap(mat.Data) == 0 || cap(a.Data) == 0 {
+		return false
+	}
+
+	off := offset(mat.Data[:1], a.Data[:1])
+
+	if off == 0 {
+		// At least one element overlaps.
+		if mat.N == a.N && mat.Stride == a.Stride {
+			panic(regionIdentity)
+		}
+		panic(regionOverlap)
+	}
+
+	if off > 0 && len(mat.Data) <= off {
+		// We know t is completely before a.
+		return false
+	}
+	if off < 0 && len(a.Data) <= -off {
+		// We know t is completely after a.
+		return false
+	}
+
+	if mat.Stride != a.Stride {
+		// Too hard, so assume the worst.
+		panic(mismatchedStrides)
+	}
+
+	// TODO(kortschak) Make this analysis more precise.
+	if off > 0 {
+		off = -off
+		mat.N, a.N = a.N, mat.N
+	}
+	if rectanglesOverlap(off, mat.N, a.N, mat.Stride) {
+		panic(regionOverlap)
+	}
+	return false
+}
+
+func (v *Vector) checkOverlap(a blas64.Vector) bool {
+	mat := v.mat
+	if cap(mat.Data) == 0 || cap(a.Data) == 0 {
+		return false
+	}
+
+	off := offset(mat.Data[:1], a.Data[:1])
+
+	if off == 0 {
+		// At least one element overlaps.
+		if mat.Inc == a.Inc && len(mat.Data) == len(a.Data) {
+			panic(regionIdentity)
+		}
+		panic(regionOverlap)
+	}
+
+	if off > 0 && len(mat.Data) <= off {
+		// We know v is completely before a.
+		return false
+	}
+	if off < 0 && len(a.Data) <= -off {
+		// We know v is completely after a.
+		return false
+	}
+
+	if mat.Inc != a.Inc {
+		// Too hard, so assume the worst.
+		panic(mismatchedStrides)
+	}
+
+	if mat.Inc == 1 || off&mat.Inc == 0 {
+		panic(regionOverlap)
+	}
+	return false
+}
+
+// rectanglesOverlap returns whether the strided rectangles a and b overlap
+// when b is offset off elements after a but has at least one element before
```

Comment:
Could you put "off" in quotes or rephrase? The documentation is really confusing when one tries to read "off" as a word instead of a variable name.